tt_metal/hw/ckernels/grayskull/common/inc/ckernel.h

```diff
@@ -267,36 +332,169 @@ inline void cfg_rmw_gpr(uint32_t cfg_addr32, uint32_t cfg_shamt, uint32_t cfg_ma
 }
 
-inline void debug_dump(uint8_t *data, uint32_t byte_size) {
-    // TODO(pk) re-implement
+inline void record_mailbox_value(uint16_t event_value) {
+  if (mailbox_index < mailbox_end) {
+    debug_mailbox_base[mailbox_index] = event_value;
+    mailbox_index++;
+  }
 }
 
-inline void tensix_sync()
-{
-    volatile uint foo = 0;
-    volatile uint *fooptr = &foo;
-    // Write to pc buffer to push all writes ahead of us.. otherwise, the pc buffer read can bypass older writes
-    pc_buf_base[1] = foo;
+inline void record_mailbox_value_with_index(uint8_t index, uint16_t event_value) {
+  if (index < mailbox_end) {
+    debug_mailbox_base[index] = event_value;
+  }
+}
 
-    // Now read -- this read will block until we're idle
-    *fooptr = pc_buf_base[1];
+// Initialize debug scratch mailbox values and range
+inline void clear_mailbox_values(uint16_t value = 0) {
+  for (int i = 0; i < mailbox_end; i++)
+    debug_mailbox_base[i] = value;
 }
 
-inline void mop_sync()
+inline uint64_t read_wall_clock()
 {
-    volatile uint foo = 0;
-    volatile uint *fooptr = &foo;
-    // Write to pc buffer to push all writes ahead of us.. otherwise, the pc buffer read can bypass older writes
-    pc_buf_base[2] = foo;
+   uint32_t timestamp_low = reg_read(RISCV_DEBUG_REG_WALL_CLOCK_L);
+   uint32_t timestamp_high = reg_read(RISCV_DEBUG_REG_WALL_CLOCK_H);
+   return ((uint64_t)timestamp_high << 32) | timestamp_low;
+}
 
-    // Now read -- this read will block until mops are done
-    *fooptr = pc_buf_base[2];
+inline void record_kernel_runtime(uint64_t kernel_runtime) {
+    debug_mailbox_base[mailbox_end - 4] = kernel_runtime;
+    debug_mailbox_base[mailbox_end - 3] = (kernel_runtime >> 16);
+    debug_mailbox_base[mailbox_end - 2] = (kernel_runtime >> 32);
+    debug_mailbox_base[mailbox_end - 1] = (kernel_runtime >> 48);
+}
+
+void debug_dump(const uint8_t *data, uint32_t byte_size);
+void debug_dump_seek(uint8_t offset);
+
+inline void stall_kernel(uint32_t num_cycles) {
+#if DELAY_EN > 0
+    TT_LLK_DUMP("stall_kernel({})", num_cycles);
+    uint32_t start_clk_l = reg_read(RISCV_DEBUG_REG_WALL_CLOCK_L);
+    uint32_t elapsed_time = 0;
+    while (elapsed_time <= num_cycles) {
+        uint32_t current_clk_l = reg_read(RISCV_DEBUG_REG_WALL_CLOCK_L);
+        if (current_clk_l >= start_clk_l) {
+            elapsed_time = current_clk_l - start_clk_l;
+        } else {
+            elapsed_time = 0xffffffff - (start_clk_l - current_clk_l);
+        }
+    }
+#endif
 }
 
+#if defined(PERF_DUMP) || DELAY_EN > 0
+extern bool record_perf_events;
+#endif
+
+// This api is inserted in the beginning of each input loop
+// Wait for all instructions of previous loop to finish before starting the next loop
+// If PERF_DUMP is enabled, always wait but only for the inputs that perf dump is enabled for
+// If PERF_DUMP is enabled, and delay is not, no need to insert these apis for unpack and math
+template<int thread_id>
+inline void serialize_input_loop_start() {
```

Comment:
The PERF_DUMP infra -- this is BBE specific?
We can have it for now in `ckernel.h` , but perhaps it can be factored out to its own file later on, so that if not used we don't have to compile it.

Reply:
PERF_DUMP is currently used by BBE, but the perf infra in the llks can be re-used by either stacks if required since it is general